### ISSUE TITLE
Auto close server session on last view close

### ISSUE THREAD

**Issue by vseryakov** [sublimelsp/LSP#869]
Hi,

This little change may be too big for a pull request. 

Working with different languages i do not want to have the dart analyst running and eating 300M+ of my memory, so it wil close the session if all dart views are closed. 

```
diff --git a/plugin/core/windows.py b/plugin/core/windows.py
index ca3f293..623e77f 100644
--- a/plugin/core/windows.py
+++ b/plugin/core/windows.py
@@ -242,6 +242,8 @@ class WindowDocumentHandler(object):
             if session.client:
                 params = {"textDocument": {"uri": filename_to_uri(file_name)}}
                 session.client.send_notification(Notification.didClose(params))
+           if not len(self._document_states):
+              session.end()

     def handle_view_saved(self, view: ViewLike) -> None:
         file_name = view.file_name()
```

**Comment by rwols** [comment]
This has been in the back of my mind for a while. It has both advantages and disadvantages. One major advantage is that we don't have to care about project switching anymore, since all old views from the previous project will be closed after switching a project. A disadvantage is that we might be starting+initializing a server quite a few times when the user opens and closes single views in a window.

**Comment by vseryakov** [comment]
right, that was a proof of concept and it works very well. i think i will extend it with a config option so it will be flexible, will do a proper pull request this time, tnx

**Comment by rwols** [comment]
I don't believe this is worth a configuration setting. It would add yet another point of maintenance.

**Comment by vseryakov** [comment]
ok, another solution, only close after 30 secs, it will fire multiple timers on every close though but i do not see how to cancel sublime timer, i guess using Python timer

```
--- windows.py.orig	2020-02-04 18:04:08.000000000 -0500
+++ windows.py	2020-02-04 18:05:40.000000000 -0500
@@ -237,11 +237,23 @@
             self._document_states.remove(file_name)
         except ValueError:
             return
+        configs = []
         for session in self._get_applicable_sessions(view, 'openClose'):
             debug('closing', file_name, session.config.name)
             if session.client:
                 params = {"textDocument": {"uri": filename_to_uri(file_name)}}
                 session.client.send_notification(Notification.didClose(params))
+                configs.append(session.config.name)
+        if not len(self._document_states):
+           self._sublime.set_timeout(lambda: self._handle_view_closed(configs), 30000)
+
+    def _handle_view_closed(self, configs: 'Any') -> None:
+        if not len(self._document_states):
+           debug('closing sessions for', configs)
+           for config_name, config_sessions in self._sessions.items():
+              for session in config_sessions:
+                if session.config.name in configs:
+                  session.end()

     def handle_view_saved(self, view: ViewLike) -> None:
         file_name = view.file_name()
```

**Comment by rwols** [comment]
This is now implemented for the ST4 branch via #1069. The session stops after three seconds after the last closed view.
